lib/Dialect/CGGI/Transforms/BooleanVectorizer.cpp

```diff
@@ -122,24 +122,19 @@ bool tryBoolVectorizeBlock(Block *block, MLIRContext &context) {
       SmallVector<StringAttr, 4> vectorizedGateOperands;
 
       for (auto *op : bucket) {
-        std::string str;
-        if (isa<cggi::AndOp>(op)) {
-          str = "and";
-        } else if (isa<cggi::NandOp>(op)) {
-          str = "nand";
-        } else if (isa<cggi::XorOp>(op)) {
-          str = "xor";
-        } else if (isa<cggi::XNorOp>(op)) {
-          str = "xnor";
-        } else if (isa<cggi::OrOp>(op)) {
-          str = "or";
-        } else if (isa<cggi::NorOp>(op)) {
-          str = "nor";
-        } else {
-          LLVM_DEBUG(llvm::dbgs()
-                     << "Try to parse boolean operation that does not exist.");
-        }
-        vectorizedGateOperands.push_back(StringAttr::get(&context, str));
+        std::string gate_str =
+            llvm::TypeSwitch<Operation *, std::string>(op)
+                .Case<cggi::AndOp>([](AndOp op) { return std::string("and"); })
+                .Case<cggi::NandOp>(
+                    [](NandOp op) { return std::string("nand"); })
+                .Case<cggi::XorOp>([](XorOp op) { return std::string("xor"); })
+                .Case<cggi::XNorOp>(
+                    [](XNorOp op) { return std::string("xnor"); })
+                .Case<cggi::OrOp>([](OrOp op) { return std::string("or"); })
+                .Case<cggi::NorOp>([](NorOp op) { return std::string("nor"); })
+                .Default([](Operation *) { return std::string("nor"); });
```

Comment:
I'm assuming that TypeSwitch required you to return a string here so for the default you returned some string - you could also return `FailureOr<std::string>` from the typeswitch and then return a `failure()` in the default case. The LLVM_DEBUG/ `llvm_unreachable` could be returned if the gate_str was a failure.

Reply:
Thanks, was looking for something like this!